File: clients/node/src/sidecar.js
'use strict'

/**
 * Binary download and subprocess management.
 * Binary is cached at ~/.cloudmeter/bin/cloudmeter-sidecar.
 * Downloaded automatically from GitHub Releases on first use.
 */

const { spawn } = require('child_process')
const { existsSync, mkdirSync, chmodSync, createWriteStream, renameSync } = require('fs')
const https = require('https')
const http  = require('http')
const os    = require('os')
const path  = require('path')

const GITHUB_REPO = process.env.CLOUDMETER_REPO || 'studymadara/cloudmeter'

function installDir() {
  return path.join(os.homedir(), '.cloudmeter', 'bin')
}

function platformAsset() {
  const plat = os.platform()
  const arch = os.arch()

  let osName, archName, ext = ''
  if      (plat === 'linux')  osName = 'linux'
  else if (plat === 'darwin') osName = 'macos'
  else if (plat === 'win32')  { osName = 'windows'; ext = '.exe' }
  else throw new Error(`[cloudmeter] Unsupported OS: ${plat}`)

  if      (arch === 'x64')   archName = 'x86_64'
  else if (arch === 'arm64') archName = 'arm64'
  else throw new Error(`[cloudmeter] Unsupported CPU architecture: ${arch}`)

  return `cloudmeter-sidecar-${osName}-${archName}${ext}`
}

function binaryPath() {
  const ext = os.platform() === 'win32' ? '.exe' : ''
  return path.join(installDir(), `cloudmeter-sidecar${ext}`)
}

function httpsGet(url) {
  return new Promise((resolve, reject) => {
    const req = https.get(url, { headers: { 'User-Agent': 'cloudmeter-node' } }, (res) => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        return httpsGet(res.headers.location).then(resolve).catch(reject)
      }
      let data = ''
      res.on('data', chunk => { data += chunk })
      res.on('end', () => resolve(data))
      res.on('error', reject)
    })
    req.on('error', reject)
  })
}

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    const follow = (u) => {
      https.get(u, { headers: { 'User-Agent': 'cloudmeter-node' } }, (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
          return follow(res.headers.location)
        }
        if (res.statusCode !== 200) {
          return reject(new Error(`[cloudmeter] HTTP ${res.statusCode} downloading binary`))
        }
        const tmp = dest + '.tmp'
        const file = createWriteStream(tmp)
        res.pipe(file)
        file.on('finish', () => {
          file.close(() => {
            renameSync(tmp, dest)
            resolve()
          })
        })
        file.on('error', reject)
      }).on('error', reject)
    }
    follow(url)
  })
}

async function ensureBinary() {
  const binary = binaryPath()
  if (existsSync(binary)) return binary

  const asset = platformAsset()
  const apiUrl = `https://api.github.com/repos/${GITHUB_REPO}/releases/latest`

  let release
  try {
    release = JSON.parse(await httpsGet(apiUrl))
  } catch (e) {
    throw new Error(`[cloudmeter] Could not fetch release info: ${e.message}`)
  }

  const found = (release.assets || []).find(a => a.name === asset)
  if (!found) {
    throw new Error(
      `[cloudmeter] No binary for your platform (${asset}). ` +
      `See https://github.com/${GITHUB_REPO}/releases`
    )
  }

  process.stderr.write(`[cloudmeter] Downloading ${asset} (${release.tag_name})...\n`)
  mkdirSync(path.dirname(binary), { recursive: true })
  await downloadFile(found.browser_download_url, binary)
  chmodSync(binary, 0o755)
  process.stderr.write(`[cloudmeter] Installed to ${binary}\n`)

  return binary
}

// ── Subprocess management ────────────────────────────────────────────────────

let _proc = null

function waitReady(port, attempts = 50) {
  return new Promise((resolve) => {
    let tries = 0
    const check = () => {
      const req = http.get(`http://127.0.0.1:${port}/api/status`, (res) => {
        if (res.statusCode === 200) return resolve()
        res.resume()
        if (++tries < attempts) setTimeout(check, 100)
        else resolve()
      })
      req.on('error', () => {
        if (++tries < attempts) setTimeout(check, 100)
        else resolve()
      })
      req.setTimeout(200, () => req.destroy())
    }
    check()
  })
}

async function start(opts = {}) {
  if (_proc && _proc.exitCode === null) return // already running

  const {
    provider                  = 'AWS',
    region                    = 'us-east-1',
    targetUsers               = 1000,
    requestsPerUserPerSecond  = 1.0,
    budgetUsd                 = 0,
    ingestPort                = 7778,
    dashboardPort             = 7777,
  } = opts

  let binary
  try {
    binary = await ensureBinary()
  } catch (e) {
    process.stderr.write(`${e.message}\n`)
    return // degrade gracefully — app still works, just no cost data
  }

  _proc = spawn(binary, [
    '--provider',                     provider,
    '--region',                       region,
    '--target-users',                 String(targetUsers),
    '--requests-per-user-per-second', String(requestsPerUserPerSecond),
    '--budget-usd',                   String(budgetUsd),
    '--ingest-port',                  String(ingestPort),
    '--dashboard-port',               String(dashboardPort),
  ], { stdio: 'ignore' })

  _proc.on('error', (e) => {
    process.stderr.write(`[cloudmeter] Sidecar process error: ${e.message}\n`)
  })

  // Register cleanup handlers once
  process.once('exit',    stop)
  process.once('SIGINT',  () => { stop(); process.exit(130) })
  process.once('SIGTERM', () => { stop(); process.exit(143) })

  await waitReady(ingestPort)
  process.stderr.write(
    `[cloudmeter] Sidecar ready — dashboard at http://localhost:${dashboardPort}\n`
  )
}

function stop() {
  if (_proc) {
    _proc.kill()
    _proc = null
  }
}

module.exports = { start, stop, ensureBinary, binaryPath }

File: clients/python/cloudmeter/_sidecar.py
"""
Binary download and subprocess management for the CloudMeter sidecar.

The binary lives at ~/.cloudmeter/bin/cloudmeter-sidecar.
On first use it is downloaded from the latest GitHub Release automatically.
"""
import atexit
import json
import os
import platform
import stat
import subprocess
import sys
import time
import urllib.error
import urllib.request
from pathlib import Path
from threading import Lock

GITHUB_REPO = os.environ.get("CLOUDMETER_REPO", "studymadara/cloudmeter")

_lock = Lock()
_process: "subprocess.Popen | None" = None
_ingest_port = 7778


def _install_dir() -> Path:
    return Path.home() / ".cloudmeter" / "bin"


def _platform_asset() -> str:
    system = platform.system().lower()
    machine = platform.machine().lower()

    if system == "linux":
        os_name = "linux"
    elif system == "darwin":
        os_name = "macos"
    elif system == "windows":
        os_name = "windows"
    else:
        raise RuntimeError(f"Unsupported OS: {system}")

    if machine in ("x86_64", "amd64"):
        arch = "x86_64"
    elif machine in ("aarch64", "arm64"):
        arch = "arm64"
    else:
        raise RuntimeError(f"Unsupported CPU architecture: {machine}")

    ext = ".exe" if system == "windows" else ""
    return f"cloudmeter-sidecar-{os_name}-{arch}{ext}"


def _binary_path() -> Path:
    ext = ".exe" if platform.system().lower() == "windows" else ""
    return _install_dir() / f"cloudmeter-sidecar{ext}"


def _download() -> None:
    asset = _platform_asset()
    binary = _binary_path()

    api_url = f"https://api.github.com/repos/{GITHUB_REPO}/releases/latest"
    try:
        req = urllib.request.Request(api_url, headers={"User-Agent": "cloudmeter-python"})
        with urllib.request.urlopen(req, timeout=10) as r:
            release = json.loads(r.read())
    except Exception as e:
        raise RuntimeError(f"[cloudmeter] Could not fetch release info: {e}") from e

    download_url = next(
        (a["browser_download_url"] for a in release.get("assets", []) if a["name"] == asset),
        None,
    )
    if download_url is None:
        raise RuntimeError(
            f"[cloudmeter] No binary found for your platform ({asset}). "
            f"See https://github.com/{GITHUB_REPO}/releases"
        )

    print(f"[cloudmeter] Downloading {asset} ({release['tag_name']})...", file=sys.stderr)
    binary.parent.mkdir(parents=True, exist_ok=True)

    tmp = binary.with_suffix(".tmp")
    try:
        urllib.request.urlretrieve(download_url, tmp)
        tmp.chmod(tmp.stat().st_mode | stat.S_IEXEC | stat.S_IXGRP | stat.S_IXOTH)
        tmp.rename(binary)
    except Exception as e:
        tmp.unlink(missing_ok=True)
        raise RuntimeError(f"[cloudmeter] Download failed: {e}") from e

    print(f"[cloudmeter] Installed to {binary}", file=sys.stderr)


def ensure_binary() -> Path:
    binary = _binary_path()
    if not binary.exists():
        _download()
    return binary


def start(
    provider: str = "AWS",
    region: str = "us-east-1",
    target_users: int = 1000,
    requests_per_user_per_second: float = 1.0,
    budget_usd: float = 0.0,
    ingest_port: int = 7778,
    dashboard_port: int = 7777,
    **_ignored,
) -> None:
    """Start the sidecar subprocess. No-op if already running."""
    global _process, _ingest_port

    with _lock:
        if _process is not None and _process.poll() is None:
            return  # already running

        try:
            binary = ensure_binary()
        except RuntimeError as e:
            print(e, file=sys.stderr)
            return  # degrade gracefully — app still runs, just no cost data

        cmd = [
            str(binary),
            "--provider", provider,
            "--region", region,
            "--target-users", str(target_users),
            "--requests-per-user-per-second", str(requests_per_user_per_second),
            "--budget-usd", str(budget_usd),
            "--ingest-port", str(ingest_port),
            "--dashboard-port", str(dashboard_port),
        ]
        _process = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        _ingest_port = ingest_port
        atexit.register(stop)

        # Wait for sidecar to be ready (up to 5 seconds)
        status_url = f"http://127.0.0.1:{ingest_port}/api/status"
        for _ in range(50):
            try:
                urllib.request.urlopen(status_url, timeout=0.5)
                print(
                    f"[cloudmeter] Sidecar ready — dashboard at http://localhost:{dashboard_port}",
                    file=sys.stderr,
                )
                return
            except (urllib.error.URLError, OSError):
                time.sleep(0.1)

        print("[cloudmeter] Warning: sidecar may not be ready yet", file=sys.stderr)


def stop() -> None:
    """Terminate the sidecar subprocess."""
    global _process
    with _lock:
        if _process is not None:
            _process.terminate()
            try:
                _process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                _process.kill()
            _process = None


def get_ingest_port() -> int:
    return _ingest_port
